src/psyclone/psyir/frontend/fparser2.py

```diff
@@ -2981,60 +3025,51 @@ def _do_construct_handler(self, node, parent):
                     raise NotImplementedError()
 
         ctrl = walk(nonlabel_do, Fortran2003.Loop_Control)
-        # do loops with no condition and do while loops
-        if not ctrl or ctrl[0].items[0]:
+        # In fparser Loop_Control has 4 children, all but one should be None,
+        # the one which is not None defines the loop boundaries style:
+        # LoopCtrl(While_Loop, Counter_Loop, Concurrent_Loop, Concurrent_Loop)
```

Comment:
I am not quite sure what the third one is maybe @rupertford can help?
The fparser rule says the fields are:
```
    F2003: While_Loop_Cntl: scalar-logical-expression, delim
    F2003: Counter_Loop_Cntl: var, lower, upper, [step], delim
    F2008: Concurrent_Loop_Cntl: conc_expr, delim
    F2018: Concurrent_Loop_Cntl: conc_expr, local_x, delim
```
but looking at the code it seems to put the "optional_delimiter" comma from the Fortran2018 R1123 rule in:
 `[ , ] CONCURRENT concurrent-header concurrent-locality`
which seems to be irrelevant for PSyIR.

I am not sure I understand this properly, but for now I updated the comment.

Reply:
If it is a concurrent loop then the arguments are [None, None, optional comment, forall_header()]. Is that what you were asking? It is all a bit convoluted to avoid changing the fparser tree for the 2003 version of loop control.